client/state/data-layer/wpcom/sites/external-contributors/index.js

```diff
@@ -0,0 +1,131 @@
+/**
+ * External Dependencies
+ */
+import { translate } from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import {
+	EXTERNAL_CONTRIBUTORS_GET_REQUEST,
+	EXTERNAL_CONTRIBUTORS_ADD_REQUEST,
+	EXTERNAL_CONTRIBUTORS_REMOVE_REQUEST,
+} from 'state/action-types';
+import { http } from 'state/data-layer/wpcom-http/actions';
+import { errorNotice } from 'state/notices/actions';
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+import {
+	receiveGetExternalContributorsSuccess,
+	receiveGetExternalContributorsFailure,
+	receiveAddExternalContributorSuccess,
+	receiveAddExternalContributorFailure,
+	receiveRemoveExternalContributorSuccess,
+	receiveRemoveExternalContributorFailure,
+} from 'state/sites/external-contributors/actions';
+
+import { registerHandlers } from 'state/data-layer/handler-registry';
+
+const handleGetExternalContributorsFailure = ( { siteId }, error ) => {
+	return [
+		errorNotice(
+			error.message
+				? translate(
+						'Failed to retrieve external contributor list for site %(siteId)d with message "%(message)s".',
+						{ args: { message: error.message, siteId } }
+				  )
+				: translate( 'Failed to retrieve external contributor list for site %(siteId)d."', {
+						args: { siteId },
+				  } )
+		),
+		receiveGetExternalContributorsFailure( siteId, error ),
+	];
+};
+
+const handleAddExternalContributorFailure = ( { siteId, userId }, error ) => {
+	return [
+		errorNotice(
+			error.message
+				? translate(
+						'Failed to add external contributor %(userId)d for site %(siteId)d with message "%(message)s".',
+						{ args: { message: error.message, siteId, userId } }
+				  )
+				: translate( 'Failed to add external contributor %(userId)d for site %(siteId)d."', {
+						args: { siteId, userId },
+				  } )
+		),
+		receiveAddExternalContributorFailure( siteId, error ),
+	];
+};
+
+const handleRemoveExternalContributorFailure = ( { siteId, userId }, error ) => {
+	return [
+		errorNotice(
+			error.message
+				? translate(
+						'Failed to remove external contributor %(userId)d for site %(siteId)d with message "%(message)s".',
+						{ args: { message: error.message, siteId, userId } }
+				  )
+				: translate( 'Failed to remove external contributor %(userId)d for site %(siteId)d."', {
+						args: { siteId, userId },
+				  } )
+		),
+		receiveRemoveExternalContributorFailure( siteId, error ),
+	];
+};
+
+registerHandlers( 'state/data-layer/wpcom/sites/external-contributors/index.js', {
+	[ EXTERNAL_CONTRIBUTORS_GET_REQUEST ]: [
+		dispatchRequest( {
+			fetch: action =>
+				http(
+					{
+						method: 'GET',
+						path: `/sites/${ action.siteId }/external-contributors`,
+						apiNamespace: 'wpcom/v2',
+					},
+					action
+				),
+			onSuccess: ( { siteId }, response ) =>
+				receiveGetExternalContributorsSuccess( siteId, response ),
+			onError: handleGetExternalContributorsFailure,
+		} ),
+	],
```

Comment:
```js
// state/data-getters.js

export const requestExternalContributors = siteId => {
	const id = `site-external-contributors-${ siteId }`;

	return requestHttpData(
		id,
		http( { method: 'GET', path: `/sites/${ siteId }/external-contributors`, apiNamespace: 'wpcom/v2' } ),
		{
			fromApi: () => ( { contributors } ) => [ [ id, contributors ] ]
		}
	);
}
```

```js
// some-component.jsx

const mapStateToProps = ( state ) => {
	const siteId = getSelectedSiteId( state );

	return {
		externalContributors: requestExternalContributors( siteId ),
	}
}
```

this is basically all we need to write, then we get network fetch state, errors, and other things for free.

Reply:
we can adapt the `POST` requests to remove the people as well, we just don't want to run it until we need to

```js
const mapDispatchToProps = ( ownProps ) => {
	remove: userId => requestExternalContributorRemoval( ownProps.siteId, userId )
}
```

and that would look like this…

// state/data-getters.js

```js
export const requestExternalContributors = ( siteId, userId ) => {
	const id = `site-external-contributors-removal-${ siteId }-${ userId }`;
	const siteId = `site-external-contributors-${ siteId }`;

	return requestHttpData(
		id,
		http( {
			method: 'POST',
			path: `/sites/${ siteId }/external-contributors/add`,
			apiNamespace: 'wpcom/v2',
			body: { user_id: userId }
		} ),
		{
			fromApi: () => () => [ [
				siteId, 
				( getHttpData( siteId ).data || [] ).filter( contributor => contributor.userId !== userId )
			] ]
		}
	);
}
```

so granted this is more complicated than the `GET` but it's still just a description of a failable HTTP call

Reply:
I wasn't aware of either of these, I will happily get rid of the boilerplate!